pkg/registry/ecr_cred_injector.go

```diff
@@ -0,0 +1,255 @@
+type ECRCredInjector struct {
+	k8sClient client.Client
+	ecrClient *ecr.Client
+	log       logr.Logger
+}
+
+func NewECRCredInjector(ctx context.Context, k8sClient client.Client, log logr.Logger) (*ECRCredInjector, error) {
+	l := log.WithName("ECRCredInjector")
+	ecrClient, err := GetECRClient(ctx, l)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ECRCredInjector{
+		k8sClient: k8sClient,
+		ecrClient: ecrClient,
+		log:       l,
+	}, nil
+}
+
+func (a *ECRCredInjector) Run(ctx context.Context) {
+	err := a.Refresh(ctx)
+	if err != nil {
+		a.log.Error(err, "Failed to inject ECR credential to docker config")
+	} else {
+		a.log.Info("ECR credential is injected to the docker config file")
+	}
+
+	for range time.Tick(time.Hour) {
+		err := a.Refresh(ctx)
+		if err != nil {
+			a.log.Error(err, "Failed to refresh ECR credential in dockerconfig file")
+		} else {
+			a.log.Info("injected ECR credential has be refreshed")
+		}
+	}
+}
+
+func (a *ECRCredInjector) Refresh(ctx context.Context) error {
+	a.log.Info("Refreshing ECR credential")
+	cred, err := GetCredential(a.ecrClient)
+	if err != nil {
+		return err
+	}
+
+	dockerSecret, err := a.GetRegistryMirrorSecret(ctx)
+	if err != nil {
+		return err
+	}
+
+	registry, err := a.GetECR(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !IsECRRegistry(registry) {
+		a.log.Info("defaultRegistry is not ECR registry, skip injecting credential to docker config")
+	}
+	// update "config.json" in dockerSecret
+	return a.InjectCredential(ctx, *dockerSecret, registry, cred)
+}
+
+// GetECR get the defaultRegistry config from package bundle controller.
+func (a *ECRCredInjector) GetECR(ctx context.Context) (string, error) {
+	pbc := &api.PackageBundleController{}
+	err := a.k8sClient.Get(ctx, types.NamespacedName{
+		Namespace: api.PackageNamespace,
+		Name:      os.Getenv("CLUSTER_NAME"),
+	}, pbc)
+	if err != nil {
+		return "", err
+	}
+
+	// defaultRegistry could be followed by path
+	ss := strings.Split(pbc.Spec.DefaultRegistry, "/")
+	return ss[0], nil
+}
+
+// GetRegistryMirrorSecret gets registry mirror secret from eksa-packages namespace
+func (a *ECRCredInjector) GetRegistryMirrorSecret(ctx context.Context) (*v1.Secret, error) {
+	var secret v1.Secret
+
+	err := a.k8sClient.Get(ctx, types.NamespacedName{
+		Namespace: api.PackageNamespace,
+		// this secret is populated by token refresher
+		Name: authenticator.MirrorCredName,
+	}, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return &secret, nil
+}
+
+// InjectCredentail update field "config.json" in the secret, which is used by packages controller's oras and helm
+func (a *ECRCredInjector) InjectCredential(ctx context.Context, secret v1.Secret, registry string, cred auth.Credential) error {
+	d := secret.Data["config.json"]
```

Comment:
do we need to verify this key exists?

Reply:
It should be ok, if this dockerconfig file doesn't exist, it would need be created so we can inject the ECR credential